### ISSUE TITLE
Can't delete folders while the app is enabled

### ISSUE THREAD

**Issue by giou** [owncloud/music#1068]
*Edit by @paulijar: Read the summary below from https://github.com/owncloud/music/issues/1068#issuecomment-1591763499*

*The original post:*

**Fixed disabling the Music app.**

```
<?php
$CONFIG = array (
  'instanceid' => '',
  'passwordsalt' => '',
  'secret' => '',
  'trusted_domains' =>
  array (
    0 => 'cloud.org',
  ),
  'datadirectory' => '/var/www/nextcloud/data',
  'dbtype' => 'mysql',
  'version' => '26.0.2.1',
  'overwrite.cli.url' => 'https://cloud.org/',
  'htaccess.RewriteBase' => '/',
  'dbname' => 'nextcloud',
  'dbhost' => '127.0.0.1',
  'dbport' => '',
  'dbtableprefix' => 'oc_',
  'mysql.utf8mb4' => true,
  'dbuser' => 'nextcloud',
  'dbpassword' => '',
  'default_phone_region' => '3166-2:ES',
  'installed' => true,
  'memcache.local' => '\\OC\\Memcache\\APCu',
  'maintenance' => false,
  'theme' => '',
  'loglevel' => 2,
  'updater.release.channel' => 'stable',
  'mail_smtpmode' => 'smtp',
  'mail_sendmailmode' => 'smtp',
  'mail_smtphost' => '',
  'mail_from_address' => '',
  'mail_domain' => '',
  'mail_smtpport' => '465',
  'mail_smtpauthtype' => 'LOGIN',
  'mail_smtpauth' => 1,
  'mail_smtpname' => '',
  'mail_smtppassword' => '',
  'mail_smtpsecure' => 'ssl',
  'updater.secret' => '',
);
```

```
[webdav] Error: TypeError: OC\Files\Cache\QuerySearchHelper::getCachesAndMountPointsForSearch(): Argument #1 ($root) must be of type OC\Files\Node\Root, OC\Files\Node\LazyRoot given, called in /var/www/nextcloud/lib/private/Files/Node/Folder.php on line 237 at <<closure>>

 0. /var/www/nextcloud/lib/private/Files/Node/Folder.php line 237
    OC\Files\Cache\QuerySearchHelper->getCachesAndMountPointsForSearch()
 1. /var/www/nextcloud/lib/private/Files/Node/Folder.php line 291
    OC\Files\Node\Folder->search()
 2. /var/www/nextcloud/apps/music/lib/Utility/Scanner.php line 383
    OC\Files\Node\Folder->searchByMime()
 3. /var/www/nextcloud/apps/music/lib/Hooks/FileHooks.php line 42
    OCA\Music\Utility\Scanner->deleteFolder()
 4. <<closure>>
    OCA\Music\Hooks\FileHooks::deleted()
 5. /var/www/nextcloud/lib/private/Hooks/EmitterTrait.php line 105
    call_user_func_array()
 6. /var/www/nextcloud/lib/private/Hooks/PublicEmitter.php line 40
    OC\Hooks\BasicEmitter->emit()
 7. /var/www/nextcloud/lib/private/Files/Node/Root.php line 143
    OC\Hooks\PublicEmitter->emit()
 8. <<closure>>
    OC\Files\Node\Root->emit()
 9. /var/www/nextcloud/lib/private/Files/Node/LazyFolder.php line 72
    call_user_func_array()
10. /var/www/nextcloud/lib/private/Files/Node/LazyFolder.php line 100
    OC\Files\Node\LazyFolder->__call()
11. /var/www/nextcloud/lib/private/Files/Node/HookConnector.php line 145
    OC\Files\Node\LazyFolder->emit()
12. /var/www/nextcloud/lib/private/legacy/OC_Hook.php line 105
    OC\Files\Node\HookConnector->delete()
13. /var/www/nextcloud/lib/private/Files/View.php line 1295
    OC_Hook::emit()
14. /var/www/nextcloud/lib/private/Files/View.php line 1165
    OC\Files\View->runHooks()
15. /var/www/nextcloud/lib/private/Files/View.php line 351
    OC\Files\View->basicOperation()
16. /var/www/nextcloud/apps/dav/lib/Connector/Sabre/Directory.php line 307
    OC\Files\View->rmdir()
17. /var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Tree.php line 179
    OCA\DAV\Connector\Sabre\Directory->delete()
18. /var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/CorePlugin.php line 281
    Sabre\DAV\Tree->delete()
19. /var/www/nextcloud/3rdparty/sabre/event/lib/WildcardEmitterTrait.php line 89
    Sabre\DAV\CorePlugin->httpDelete()
20. /var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php line 472
    Sabre\DAV\Server->emit()
21. /var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php line 253
    Sabre\DAV\Server->invokeMethod()
22. /var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php line 321
    Sabre\DAV\Server->start()
23. /var/www/nextcloud/apps/dav/lib/Server.php line 366
    Sabre\DAV\Server->exec()
24. /var/www/nextcloud/apps/dav/appinfo/v2/remote.php line 35
    OCA\DAV\Server->exec()
25. /var/www/nextcloud/remote.php line 172
    require_once("/var/www/nextcl ... p")

DELETE /remote.php/dav/files/giou/Obsidiane
from IP by giou at 2023-05-26T21:05:08+00:00
```

**Comment by rcjcarr** [comment]
Have the same issue, [https://help.nextcloud.com/t/error-deleting-folders-or-moving-folders/162994](https://help.nextcloud.com/t/error-deleting-folders-or-moving-folders/162994)

**Comment by paulijar** [comment]
Looks like a Nextcloud bug to me and likely caused by this commit: https://github.com/nextcloud/server/commit/dbfd2f936aaae95aabd52ac8de03b48a813bbb8f. Maybe @blizzz could comment?

In that commit, this new function assumes the first argument to be of type `Root`:
https://github.com/nextcloud/server/blob/dbfd2f936aaae95aabd52ac8de03b48a813bbb8f/lib/private/Files/Cache/QuerySearchHelper.php#L202

And it's called here by passing `$this->root` as the first argument:
https://github.com/nextcloud/server/blob/dbfd2f936aaae95aabd52ac8de03b48a813bbb8f/lib/private/Files/Node/Folder.php#L238

But `$this->root` is not guaranteed to be of the type `Root`. From the constructor argument list, we only know that it implements the interface `IRootFolder` and even that is not enforced by any type declaration:
https://github.com/nextcloud/server/blob/dbfd2f936aaae95aabd52ac8de03b48a813bbb8f/lib/private/Files/Node/Node.php#L69-L75

**Comment by paulijar** [comment]
As a quick-and-dirty monkey patch, you may modify the file `lib/private/Files/Cache/QuerySearchHelper.php` in your installation. Locate the line
`public function getCachesAndMountPointsForSearch(Root $root, string $path, bool $limitToHome = false): array {`
and modify it to be
`public function getCachesAndMountPointsForSearch($root, string $path, bool $limitToHome = false): array {`
i.e. remove the type declaration `Root`.

**Comment by blizzz** [comment]
Thanks for the report and investigation, and pinging me. I'll try to have a look into it this week.

**Comment by AubinF** [comment]
Same issue here. Thanks @blizzz @paulijar for your responsiveness 💪
@paulijar 's quick and dirty patch works for me.

**Comment by JoshuaPettus** [comment]
Fun story, Just today I accidentally deleted a folder in the NC UI and it failed.  Had no idea why but just shrugged it off and called it good luck.  Now I stumbled on this!

**Comment by Kaliumhexacyanoferrat** [comment]
Can confirm disabling the music app allows to delete files in nextcloud once again.

**Comment by tryallthethings** [comment]
Can confirm this as well. Disabling the music app is a workaround.

**Comment by adripo** [comment]
Same problem for me with Nextcloud 26. I had to disable the app. Please have a look at this issue as it is a production bug and should be fixed ASAP. Thanks

**Comment by JoshuaPettus** [comment]
Guys, read through the thread, it's a Nextcloud server issue that Music happens to be triggering.  Blizzz put has been working on a patch

**Comment by paulijar** [comment]
Yeah, so for those of your who don't "have time" to read the whole thread: There is a bug in Nextcloud versions 25.0.7, 26.0.2, and 27.0.0 which happens to trigger when the Music app reacts on the folder deletion. There may be other occasions too, but I'm not aware of those. There is already a [PR](https://github.com/nextcloud/server/pull/38625) fixing the issue but it's still waiting for some reviews before it can be merged. I hope this will be fixed in Nextcloud version 25.0.8, 26.0.3, and 27.0.1. Until then, your options are:
- not to use Music
- use an older version of Nextcloud
- implement the easy monkey patch described in https://github.com/owncloud/music/issues/1068#issuecomment-1566244178

Commenting "me too" here doesn't help in the official fix to appear any sooner.

**Comment by JoshuaPettus** [comment]
4th option is to simply do your folder deletion on the desktop app instead of the web UI.

**Comment by adripo** [comment]
@JoshuaPettus sorry for being abrupt and bothering you with my last message. I read the thread but didn't pay enough attention to understand that you were talking about an issue on the server. I was having different issues on my instance after the last update. For the moment I disabled Music app.

Also, thank you @paulijar for the summary

**Comment by meschg** [comment]
Hello everyone, 
just had the same issue on 27.0.0
I cloud not move or delete one folder and after disabling music 1.8.4 it work instantly.
This affected in browser and also the sync client. The folder was not allowed to sync and was reverted on the client.
Kind regards,

**Comment by blizzz** [comment]
@meschg please please please read the thread before commenting. https://github.com/owncloud/music/issues/1068#issuecomment-1591763499

**Comment by blizzz** [comment]
> I hope this will be fixed in Nextcloud version 25.0.8, 26.0.3, and 27.0.1. Until then, your options are:

Unfortunately it will be only 25.0.9 and 26.0.4 (and 27.0.1) :(

**Comment by gitdode** [comment]
Make an update - discover an issue - check the web - find an easy workaround - issue addressed already - release with fix coming up soon... great product & community!

**Comment by EpicJosch** [comment]
I have the same problem that i cant delete files/folders with music app enabled. Is this scheduled to be fixed in the next versions?

**Comment by paulijar** [comment]
@EpicJosch Please see https://github.com/owncloud/music/issues/1068#issuecomment-1591763499 for the summary and status of this issue. To add to that, the fix is now on track to be included in the next point releases of NC26 and NC27. NC25 apparently still needs some manual work, as the automatic backporting has failed.

**Comment by blizzz** [comment]
Backports are at https://github.com/nextcloud/server/pull/38976 and https://github.com/nextcloud/server/pull/38977 and ready to review. That one for 25 will follow based on the 26 one once it is merged.

**Comment by gitdode** [comment]
@cyberpower678 You could apply this [workaround](https://github.com/owncloud/music/issues/1068#issuecomment-1566244178) also in your container. Of course it won't survive a container recreation but that won't happen so frequently I guess?
I can confirm it works well for me as container.

**Comment by cyberpower678** [comment]
> @cyberpower678 You could apply this [workaround](https://github.com/owncloud/music/issues/1068#issuecomment-1566244178) also in your container. Of course it won't survive a container recreation but that won't happen so frequently I guess?
> I can confirm it works well for me as container.

You mean you don’t need to restart the web server?   Actually now that I think about it, that wouldn’t make sense.  But since the container does get redeployed more than I like, I’ll keep the app off.

**Comment by JoshuaPettus** [comment]
I can confirm it's working in 27.0.1

**Comment by paulijar** [comment]
As the problem has now been fixed by Nextcloud in all the affected versions (NC25, NC26, NC27), I'm closing this issue. Disclaimer: I didn't test the fixes myself, but I can see that the fix PR is included in each of the latest releases 25.0.9, 26.0.4, and 27.0.1.
